[Intelligence Effects] Add a platform abstraction layer to be able to use intelligence effects in a platform-agnostic manner

rr-codes opened the pull request:
#### d03f052253e8b2d8eaad250db4245cb898106220
<pre>
[Intelligence Effects] Add a platform abstraction layer to be able to use intelligence effects in a platform-agnostic manner
<a href="https://bugs.webkit.org/show_bug.cgi?id=281801">https://bugs.webkit.org/show_bug.cgi?id=281801</a>
<a href="https://rdar.apple.com/138225697">rdar://138225697</a>

Reviewed by Abrar Rahman Protyasha.

The interface for creating intelligence effects on macOS compared to iOS is quite different. To make
this difference easier to manage, a high level abstraction type is created which mostly abstracts away
all the platform-specific parts so that it may be consumed without the client having to worry about
making any logic changes platform-dependent.

The abstraction layer&apos;s interface attempts to be as similar to iOS as possible, since abstracting the
macOS interface in terms of iOS was more elegant.

This will soon be adopted and used by an upcoming intelligence effects coordinator type.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift: Added.
(PlatformIntelligenceTextEffect.chunk):
(PlatformIntelligenceTextEffectViewSource.textPreview(for:)):
(PlatformIntelligenceTextEffectViewSource.updateTextChunkVisibility(_:visible:)):
(PlatformIntelligenceTextEffectViewSource.performReplacementAndGeneratePreview(for:effect:animation:)):
(PlatformIntelligenceTextEffectViewSource.replacementEffectWillBegin(_:)):
(PlatformIntelligenceTextEffectViewSource.replacementEffectDidComplete(_:)):
(UITextEffectViewSourceAdapter.wrapped):
(UITextEffectViewSourceAdapter.targetedPreview(for:)):
(UITextEffectViewSourceAdapter.updateTextChunkVisibilityForAnimation(_:visible:)):
(UIReplacementTextEffectDelegateAdapter.view):
(UIReplacementTextEffectDelegateAdapter.replacementEffectDidComplete(_:)):
(UIReplacementTextEffectDelegateAdapter.performReplacementAndGeneratePreview(for:effect:animation:)):
(WTTextPreviewAsyncSourceAdapter.textPreviews(for:)):
(WTTextPreviewAsyncSourceAdapter.textPreview(for:)):
(WTTextPreviewAsyncSourceAdapter.updateIsTextVisible(_:for:)):
(PlatformIntelligenceTextEffectView.wrappedEffectIDToPlatformEffects):
(PlatformIntelligenceTextEffectView.platformEffectIDToWrappedEffectIDs):
(PlatformIntelligenceTextEffectView.updateFrame):
(PlatformIntelligenceTextEffectView.removeEffect(_:)):
(PlatformIntelligenceTextEffectView.removeAllEffects):

Canonical link: <a href="https://commits.webkit.org/285512@main">https://commits.webkit.org/285512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36a740954ac30fd61ac6e59911ae7fbcbfdcb521

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72815 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77008 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24049 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74930 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23865 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57250 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15745 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75882 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47234 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37679 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43881 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20138 "Found 1 new test failure: imported/w3c/web-platform-tests/IndexedDB/idbobjectstore_keyPath.any.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/22378 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65734 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78684 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17061 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19639 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65700 "Found 5 new test failures: imported/blink/compositing/animation/hidden-animated-layer-should-not-have-scrollbars.html imported/w3c/web-platform-tests/css/css-animations/display-none-to-display-block.html imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-animation-fallback-missing-100-percent.html imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-animation-non-zero-size-element-change-to-zero.html imported/w3c/web-platform-tests/css/css-backgrounds/animations/background-color-animation-with-table2.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17109 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62680 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64977 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13289 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6942 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11205 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48038 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2825 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49105 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50400 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48850 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
```diff
@@ -0,0 +1,485 @@
+// MARK: Platform abstraction type aliases
+
+#if canImport(AppKit)
+typealias PlatformView = NSView
```
Two questions...

1. I'm guessing the C++ type aliases defined in `PlatformView.h` can't be reused?
2. Have we thought about setting up some platform abstractions at large for WebKitSwift? Would a `WebKitSwift/platform/cocoa/PlatformView.swift` file make sense with this content make sense? This suggestion may be completely off-key so feel free to ignore.

```swift
// PlatformView.swift

#if canImport(AppKit)
typealias PlatformView = NSView
#else
typealias PlatformView = UIView
#endif
```

rr-codes replied:
> I'm guessing the C++ type aliases defined in PlatformView.h can't be reused?
nope

> Have we thought about setting up some platform abstractions at large for WebKitSwift? Would a WebKitSwift/platform/cocoa/PlatformView.swift file make sense with this content make sense? This suggestion may be completely off-key so feel free to ignore.

This would probably be worth doing if WebKitSwift is going to be long lived, but since we’re trying to migrate to being able to write Swift in WebKit, doing this now would make such a migration a bit more painful

aprotyas replied:
Oh that's good to know, maybe we can have `PlatformView.swift` live next to `PlatformView.h` when we can _just_ write Swift in WebKit.

aprotyas commented on Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
```diff
@@ -0,0 +1,485 @@
+@MainActor protocol PlatformIntelligenceTextEffect: Equatable, Identifiable where ID == PlatformIntelligenceTextEffectID {
+    associatedtype C: PlatformIntelligenceTextEffectChunk
+
+    var chunk: C { get }
```
Maybe `associatedtype Chunk: PlatformIntelligenceTextEffectChunk`? Just to keep later references like `Source.Chunk == Chunk` easy to track.

rr-codes replied:
sure; I mainly went with C because I really don’t like the word “chunk” in general lol

aprotyas commented on Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift:
```diff
@@ -0,0 +1,485 @@
+        view.wrappedEffectIDToPlatformEffects[wrappedEffect.id] = self
+        view.platformEffectIDToWrappedEffectIDs[self.id] = wrappedEffect.id
+#else
+        // WritingToolsUI usually performs a replacement effect using the following flow:
+        //
+        //  1. A source effect is created
+        //  2. WTUI then requests a text preview of the source text.
+        //  3. WTUI requests the relevant text to become invisible.
+        //  4. WTUI invokes the 'pre-completion' block associated with the source effect. This gets invokes after the preview is generated
+        //     but before the effect actually begins. In this block, it is intended that the client performs the replacement, and then
+        //     begins a destination effect.
+        //  5. After adding the destination effect, WTUI requests a text preview of the now-replaced text.
+        //  6. WTUI then uses this text preview to begin the destination effect.
+        //  7. This results in the destination effect beginning immediately after the source effect begins, with no delay.
+        //
+        // However, if the replacement must happen asynchronously, the destination effect cannot begin immediately after the source effect,
+        // since it needs to wait for the text to be replaced so that it can then request a preview of the replaced text. Consequently, the
+        // whole replacement effect will not look correct, since the destination effect will lag behind the source effect.
+        //
+        // To address this gap in the interface of WTUI, the source and destination effect are abstracted into a single replacement effect.
+        // When adding this replacement effect, the replacement happens prior to both the source and destination effects starting. Specifically,
+        // this effectively works the way the UIKit interface does things. This allows the destination effect to happen immediately when needed,
+        // since the preview generation and replacement has already happened.
+
+        // First, a text preview of the source text is manually generated and saved.
+        // When WTUI requests the text preview of the source, this preview can be synchronously returned in the delegate method rather than
+        // a new preview being generated on-demand.
```
Woah

rr-codes committed:
[Intelligence Effects] Add a platform abstraction layer to be able to use intelligence effects in a platform-agnostic manner
https://bugs.webkit.org/show_bug.cgi?id=281801
rdar://138225697

Reviewed by Abrar Rahman Protyasha.

The interface for creating intelligence effects on macOS compared to iOS is quite different. To make
this difference easier to manage, a high level abstraction type is created which mostly abstracts away
all the platform-specific parts so that it may be consumed without the client having to worry about
making any logic changes platform-dependent.

The abstraction layer's interface attempts to be as similar to iOS as possible, since abstracting the
macOS interface in terms of iOS was more elegant.

This will soon be adopted and used by an upcoming intelligence effects coordinator type.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WritingTools/PlatformIntelligenceTextEffectView.swift: Added.
(PlatformIntelligenceTextEffect.chunk):
(PlatformIntelligenceTextEffectViewSource.textPreview(for:)):
(PlatformIntelligenceTextEffectViewSource.updateTextChunkVisibility(_:visible:)):
(PlatformIntelligenceTextEffectViewSource.performReplacementAndGeneratePreview(for:effect:animation:)):
(PlatformIntelligenceTextEffectViewSource.replacementEffectWillBegin(_:)):
(PlatformIntelligenceTextEffectViewSource.replacementEffectDidComplete(_:)):
(UITextEffectViewSourceAdapter.wrapped):
(UITextEffectViewSourceAdapter.targetedPreview(for:)):
(UITextEffectViewSourceAdapter.updateTextChunkVisibilityForAnimation(_:visible:)):
(UIReplacementTextEffectDelegateAdapter.view):
(UIReplacementTextEffectDelegateAdapter.replacementEffectDidComplete(_:)):
(UIReplacementTextEffectDelegateAdapter.performReplacementAndGeneratePreview(for:effect:animation:)):
(WTTextPreviewAsyncSourceAdapter.textPreviews(for:)):
(WTTextPreviewAsyncSourceAdapter.textPreview(for:)):
(WTTextPreviewAsyncSourceAdapter.updateIsTextVisible(_:for:)):
(PlatformIntelligenceTextEffectView.wrappedEffectIDToPlatformEffects):
(PlatformIntelligenceTextEffectView.platformEffectIDToWrappedEffectIDs):
(PlatformIntelligenceTextEffectView.updateFrame):
(PlatformIntelligenceTextEffectView.removeEffect(_:)):
(PlatformIntelligenceTextEffectView.removeAllEffects):

Canonical link: https://commits.webkit.org/285512@main
